DEX mini tradeform, wallets, order history page

song50119 opened the pull request:
Work towards #619

- Add mini trade form
- Submit Market Order buy or sell DCR
- Add wallets manage page and order history page
- Display wallet address to deposit

Commit:
Add mini trade form
- Add BTC wallet

Commit:
Add Dex wallets page
- Display wallet status sync, locked, open, balance on the UI
- Create deposit button to show address on modal, can copy or scan QR code that address
- Buttons lock and unlock to handler wallet
- Create listener to refresh wallet state on the UI

Commit:
Create order history page
- Add listener for order history page to update order state

Commit:
Create dex connection with websocket server
- Initialize websocket connection and subscribe to market after login or registration
- Read messages source type Notification from dex connnection
- Handler book_order route add new orderBook to Market Page and refresh UI.
- Handler unbook_order route remove orderBook from Market Page and refresh UI

Commit:
Fix lint error

Commit:
Add withdraw modal layout

Commit:
Fix crash when login Dex

Commit:
Add settings page
- Add settings button to market page to be able to navigate to settings page
- Display account info on setting page
- Implement disable account, change app password function
- Ability export account info to json file
- Ability import account from json file

itswisdomagain commented on ui/load/load.go:
```diff
@@ -38,7 +38,7 @@ type Icons struct {
 	NavigationArrowForward, ActionCheck, ChevronRight, NavigationCancel, NavMoreIcon,
 	ImageBrightness1, ContentClear, DropDownIcon, Cached, ContentRemove, ConcealIcon, RevealIcon,
-	SearchIcon *widget.Icon
+	SearchIcon, ActionSwapHoriz *widget.Icon
```
`ActionSwapHoriz` doesn't clear tell what this icon is. I recommend `SwapIcon` or `ExchangeIcon`.

itswisdomagain commented on ui/page/dexclient/markets.go:
```diff
@@ -42,10 +54,14 @@ func (pg *Page) dexMarket(mktName string) *core.Market {
+		initialize:    l.Theme.Button("Start using now"),
+		addDex:        l.Theme.Button("Add a dex"),
+		sync:          l.Theme.Button("Start sync to continue"),
+		toWallets:     l.Theme.Button("Settings Wallets"),
```
`toWallets` should be `walletSettingsButton`
"Settings Wallets" should be "Wallet Settings"

song50119 replied:
Yh, walletSettings is ok. don't need Button at the end

itswisdomagain commented on ui/page/dexclient/markets.go:
```diff
@@ -79,13 +101,50 @@ func (pg *Page) Layout(gtx C) D {
+				return D{}
+			}
+
+			if dex.PendingFee == nil && pg.miniTradeFormWdg != nil {
```
Is it possible for `dex.PendingFee` to be nil and `pg.miniTradeFormWdg` is also nil? If not, `if dex.PendingFee == nil {` is enough condition.

I'd probably even go with `if dex.PendingFee != nil { show registrationStatusLayout }`. Then show `miniTradeFormWdg` lastly.

itswisdomagain commented on ui/page/dexclient/markets.go:
```diff
@@ -79,13 +101,50 @@ func (pg *Page) Layout(gtx C) D {
-			return pg.pageSections(gtx, pg.registrationStatusLayout)
+			dex := pg.dex()
+			if !dex.Connected {
+				return D{}
```
Show a message? An empty screen doesn't tell the user anything.

itswisdomagain commented on ui/page/dexclient/markets.go:
```diff
@@ -79,13 +101,50 @@ func (pg *Page) Layout(gtx C) D {
 	return components.UniformPadding(gtx, body)
 }
 
+func (pg *Page) settingsLayout(gtx C) D {
```
The `registrationStatusLayout` is referred to first (line 120) before the `settingsLayout`. Re-order the methods to position the `registrationStatusLayout` before the `settingsLayout`. It's easier to follow the code when methods are defined close to where they are called.

itswisdomagain commented on ui/page/dexclient/markets.go:
```diff
@@ -216,8 +355,60 @@ func (pg *Page) readNotifications() {
+}
+
+func (pg *Page) listenerMessages(host string) {
+	msgs := pg.Load.MessageSource(host)
```
You don't need to open a websocket connection to the DEX to listen for messages, that's handled internally by Core. To know when there is an update to the order book, note that the `pg.Dexc().Core().SyncBook` method returns a `core.BookFeed` object that has a `Next()` method for receiving updates.  Not completely familiar with it but it should work for your use case. The basic idea is:

```
bookFeed, err := pg.Dexc().Core().SyncBook(nd.Host, mkt.BaseID, mkt.QuoteID)
for {
      <- bookFeed.Next()
      // Update received, reload the book
      pg.getOrderBook(nd.Host, mkt.BaseID, mkt.QuoteID)
}
```

itswisdomagain reviewed (changes requested):
Preliminary review, will take a further look later. 
Please include in your PR description that this PR depends on #676 and the first 15 commits here are from #676. I only reviewed the commits after the 15th commit.

Commit:
Remove dex connection from Load
- Use Core().SyncBook then read messages from (core.BookFeed).Next() channel to refresh UI
- After register dex, reload to Market Page to listen messages
- Display UI text when dex is not connected, refresh window when dex reconnect

Sirmorrison commented on ui/page/dexclient/markets.go:
```diff
@@ -151,16 +182,45 @@ func (pg *Page) registrationStatusLayout(gtx C) D {
+					return layout.Inset{Left: values.MarginPadding10}.Layout(gtx, func(gtx C) D {
+						return pg.walletSettings.Layout(gtx)
+					})
+				}),
```
```suggestion
			return layout.Flex{}.Layout(gtx,
				layout.Rigid(pg.ordersHistory.Layout),
				layout.Rigid(func(gtx C) D {
					return layout.Inset{Left: values.MarginPadding10}.Layout(gtx, pg.walletSettings.Layout)
				}),
```

Sirmorrison commented on ui/page/dexclient/markets.go:
```diff
@@ -151,16 +182,45 @@ func (pg *Page) registrationStatusLayout(gtx C) D {
+	})
+}
+
+func (pg *Page) dexNotConnectLabel(host string) decredmaterial.Label {
```
```suggestion
func (pg *Page) dexConnectionErrorLabel(host string) decredmaterial.Label {
```

Sirmorrison commented on ui/page/dexclient/mini_trade_form.go:
```diff
@@ -0,0 +1,205 @@
+					m.orderedAmount.Editor.SetText(value)
+				}
+			}
+		}
```
```suggestion
			if m.invoicedAmount.Editor.Focused() {
				_, change := decredmaterial.HandleEditorEvents(m.invoicedAmount.Editor)
				if change {
					value := m.suggestValue(ord, m.invoicedAmount.Editor.Text(), m.isSell, true)
					m.orderedAmount.Editor.SetText(value)
				}
			}
```

Sirmorrison commented on ui/page/dexclient/mini_trade_form.go:
```diff
@@ -0,0 +1,205 @@
+				}
+			}
+		}
+	}
```
```suggestion
			if m.orderedAmount.Editor.Focused() {
				_, change := decredmaterial.HandleEditorEvents(m.orderedAmount.Editor)
				if change{
					value := m.suggestValue(ord, m.orderedAmount.Editor.Text(), m.isSell, false)
					m.invoicedAmount.Editor.SetText(value)
				}
			}
		}
```

Sirmorrison commented on ui/page/dexclient/orders_history_page.go:
```diff
@@ -0,0 +1,191 @@
+		gtx.Constraints.Min.X = gtx.Constraints.Max.X
+		return layout.UniformInset(values.MarginPadding16).Layout(gtx, body)
+	})
+}
```
This is used only once in line 66. I dont think we need it as it is not many codes.

Sirmorrison commented on ui/page/dexclient/orders_history_page.go:
```diff
@@ -0,0 +1,191 @@
+									}),
+									layout.Flexed(0.125, func(gtx C) D {
+										return pg.Theme.Label(values.TextSize14, "Time").Layout(gtx)
+									}),
```
```suggestion
									layout.Flexed(0.125, pg.layoutRows("Trade")),
									layout.Flexed(0.125, pg.layoutRows("Side")), 
									layout.Flexed(0.125, pg.layoutRows("Rate")), 
									layout.Flexed(0.125, pg.layoutRows("Quantity")),
									layout.Flexed(0.125, pg.layoutRows("Filled")), 
									layout.Flexed(0.125, pg.layoutRows("Settled")),
									layout.Flexed(0.125, pg.layoutRows("Status")),
									layout.Flexed(0.125, pg.layoutRows("Time")),
```

Sirmorrison commented on ui/page/dexclient/orders_history_page.go:
```diff
@@ -0,0 +1,191 @@
+										}),
+										layout.Flexed(0.125, func(gtx C) D {
+											return pg.Theme.Label(values.TextSize14, timeSince(ord.Stamp)).Layout(gtx)
+										}),
```
```suggestion
layout.Flexed(0.125, pg.layoutRows(typeString(ord))),
										layout.Flexed(0.125, pg.layoutRows(sellString(ord))),
										layout.Flexed(0.125, pg.layoutRows(rateString(ord))),
										layout.Flexed(0.125, pg.layoutRows(formatCoinValue(ord.Qty))), 
										layout.Flexed(0.125, pg.layoutRows(fmt.Sprintf("%.1f%%", (float64(ord.Filled)/float64(ord.Qty))*100))),
										layout.Flexed(0.125, pg.layoutRows(fmt.Sprintf("%.1f%%", settled(ord)/float64(ord.Qty)*100))),
										layout.Flexed(0.125, pg.layoutRows(statusString(ord))),
										layout.Flexed(0.125, pg.layoutRows(timeSince(ord.Stamp))), 
```

Sirmorrison commented on ui/page/dexclient/orders_history_page.go:
```diff
@@ -0,0 +1,191 @@
+					Left:  values.MarginPadding10,
+					Right: values.MarginPadding10,
+				}.Layout(gtx, func(gtx C) D {
+					return pg.pageSections(gtx, func(gtx C) D {
```
This can go.

Sirmorrison commented on ui/page/dexclient/orders_history_page.go:
```diff
@@ -0,0 +1,191 @@
+									layout.Flexed(0.125, func(gtx C) D {
+										return pg.Theme.Label(values.TextSize14, "Rate").Layout(gtx)
+									}),
+									layout.Flexed(0.125, func(gtx C) D {
```
Instead of having to repeat same label multiple time, you can add this to a function.
```
func (pg *OrdersHistoryPage) layoutRows(text string) layout.Widget {
	return pg.Theme.Label(values.TextSize14, text).Layout
}
```

itswisdomagain replied:
Or `layoutLabel`. It really is laying out a label, although the caller in this instance is using the label as a row, which is fine, but the method lays out a label.

Sirmorrison commented on ui/page/dexclient/orders_history_page.go:
```diff
@@ -0,0 +1,191 @@
+							}),
+							layout.Rigid(func(gtx C) D {
+								return pg.Theme.List(pg.list).Layout(gtx, len(pg.orders), func(gtx C, i int) D {
+									ord := pg.orders[i]
```
ord does not seem right.
```suggestion
									order := pg.orders[i]
```

Sirmorrison commented on ui/page/dexclient/orders_history_page.go:
```diff
@@ -0,0 +1,191 @@
+		case n := <-ch:
+			switch n.Type() {
+			case core.NoteTypeOrder:
+				ord := n.(*core.OrderNote)
```
kindly rename this.. `coreOrder`

Sirmorrison commented on ui/page/dexclient/wallet_deposit_modal.go:
```diff
@@ -0,0 +1,230 @@
+			return layout.Flex{Alignment: layout.Middle}.Layout(gtx,
+				layout.Rigid(func(gtx C) D {
+					return md.Load.Theme.Label(values.TextSize20, "Deposit").Layout(gtx)
+				}),
```
```suggestion
				layout.Rigid( md.Load.Theme.Label(values.TextSize20, "Deposit").Layout),
```

Sirmorrison commented on ui/page/dexclient/wallet_deposit_modal.go:
```diff
@@ -0,0 +1,230 @@
+				}),
+				layout.Rigid(func(gtx C) D {
+					return md.Load.Theme.Label(values.TextSize20, md.walletInfoWidget.coinName).Layout(gtx)
+				}),
```
```suggestion
				layout.Rigid( md.Load.Theme.Label(values.TextSize20, md.walletInfoWidget.coinName).Layout),
```

Sirmorrison commented on ui/page/dexclient/wallet_deposit_modal.go:
```diff
@@ -0,0 +1,230 @@
+					}),
+					layout.Rigid(func(gtx C) D {
+						return md.newAddr.Layout(gtx)
+					}),
```
```suggestion
					layout.Rigid(md.cancel.Layout),
					layout.Rigid(md.copy.Layout),
					layout.Rigid(md.newAddr.Layout),
```

Sirmorrison commented on ui/page/dexclient/wallet_deposit_modal.go:
```diff
@@ -0,0 +1,230 @@
+						Right:  values.MarginPadding16,
+					}.Layout(gtx, func(gtx layout.Context) layout.Dimensions {
+						return md.Theme.Label(values.TextSize14, md.wallAddress).Layout(gtx)
+					})
```
```suggestion
					}.Layout(gtx,  md.Theme.Label(values.TextSize14, md.wallAddress).Layout)
```

Sirmorrison commented on ui/page/dexclient/wallet_withdraw_modal.go:
```diff
@@ -0,0 +1,189 @@
+			return layout.Flex{Alignment: layout.Middle}.Layout(gtx,
+				layout.Rigid(func(gtx C) D {
+					return md.Load.Theme.Label(values.TextSize20, "Withdraw").Layout(gtx)
+				}),
```
```suggestion
				layout.Rigid(md.Load.Theme.Label(values.TextSize20, "Withdraw").Layout),
```

Sirmorrison commented on ui/page/dexclient/wallet_withdraw_modal.go:
```diff
@@ -0,0 +1,189 @@
+				}),
+				layout.Rigid(func(gtx C) D {
+					return md.Load.Theme.Label(values.TextSize20, strings.ToUpper(md.walletInfoWidget.coinName)).Layout(gtx)
+				}),
```
```suggestion
				layout.Rigid(md.Load.Theme.Label(values.TextSize20, strings.ToUpper(md.walletInfoWidget.coinName)).Layout),
```

Sirmorrison commented on ui/page/dexclient/wallets_page.go:
```diff
@@ -0,0 +1,354 @@
+											Right: values.MarginPadding8,
+										}.Layout(gtx, func(gtx C) D {
+											return pg.Theme.Body2(asset.Info.Name).Layout(gtx)
+										})
```
```suggestion
										}.Layout(gtx, pg.Theme.Body2(asset.Info.Name).Layout)
```

Sirmorrison commented on ui/page/dexclient/wallets_page.go:
```diff
@@ -0,0 +1,354 @@
+									}),
+									layout.Rigid(func(gtx C) D {
+										return pg.Theme.Body2(fmt.Sprintf("(%s)", strings.ToUpper(asset.Symbol))).Layout(gtx)
+									}),
```
```suggestion
									layout.Rigid(pg.Theme.Body2(fmt.Sprintf("(%s)", strings.ToUpper(asset.Symbol))).Layout),
```

Sirmorrison commented on ui/page/dexclient/wallets_page.go:
```diff
@@ -0,0 +1,354 @@
+									return b.Layout(gtx, func(gtx C) D {
+										return inset.Layout(gtx, func(gtx C) D {
+											return pg.Theme.Label(values.MarginPadding12, label).Layout(gtx)
+										})
```
```suggestion
										return inset.Layout(gtx,  pg.Theme.Label(values.MarginPadding12, label).Layout)
```

itswisdomagain commented on ui/page/dexclient/markets.go:
```diff
@@ -88,7 +96,32 @@ func (pg *Page) Layout(gtx C) D {
+
+			mkt := pg.dexMarket(dex, marketIDSelected)
+			if pg.miniTradeFormWdg == nil {
+				// TODO: renew miniTradeFormWdg if change host or market
```
Can take care of this by adding `pg.miniTradeFormWdg.host != dex.Host || pg.miniTradeFormWdg.mkt.Name != mkt.Name` to the if condition.

itswisdomagain commented on ui/page/dexclient/markets.go:
```diff
@@ -128,12 +161,10 @@ func (pg *Page) dex() *core.Exchange {
 	if dex.PendingFee == nil {
-		// TODO: render trade UI
 		return pg.Theme.Label(values.TextSize14, "Registration fee payment successful!").Layout(gtx)
 	}
```
The caller of this method already checks these conditions, no longer need them here, as long as the `dex` here is the same `dex`  that was checked before this method was called. To ensure that it is the same dex, pass that dex into this method instead of getting another reference to the dex using `pg.dex()`.

itswisdomagain commented on ui/page/dexclient/markets.go:
```diff
@@ -88,7 +96,32 @@ func (pg *Page) Layout(gtx C) D {
+				)
+			}
+
+			mkt := pg.dexMarket(dex, marketIDSelected)
```
You don't need the `dexMarket` method anymore, should just do `dex.Markets[marketIDSelected]` here. On that note, rename `marketIDSelected` to `defaultMarketID` and add a comment that explains that the actual market ID to use should be selected by the user, not hard-coded.

itswisdomagain commented on ui/page/dexclient/markets.go:
```diff
@@ -24,18 +24,24 @@ const MarketPageID = "Markets"
+	sync             decredmaterial.Button
+	miniTradeFormWdg *miniTradeFormWidget
+	walletSettings   decredmaterial.Button
+	ordersHistory    decredmaterial.Button
```
Please add `Btn` to the end of the variables that are buttons, just as the `miniTradeFormWdg` variable ends with a `Wdg`.

itswisdomagain reviewed (changes requested):
Stopping my review here for now to request an important feature for this PR, if not already implemented. Does this PR allow a user select the market they want to trade on? I think it's important even though users can currently only trade dcr-btc markets; the market ID may not be dcr_btc but btc_dcr. Also, it's generally a good idea for users to see the markets supported by a DEX.

Commit:
Add Selector Dex Modal
- Add prefix Btn for fields have type of decredmaterial.Button, or decredmaterial.Clickable
- Save dex host to db to be able to persist dex state when reopen app

Commit:
Refactor Add Dex Modal
- Add Dex Modal, select cert file to register instead of input a string
- Market Page select first market in the list of dex.Markets as default instead of hardcode value dcr_btc

Commit:
Use (*core.Core).DiscoverAccount for registration
- When add new dex, use (*core.Core).DiscoverAccount instead (*dcrlibwallet.DexClient).DEXServerInfo, DiscoverAccount have account paid status returned then can fix the bug when disable account and add again
- Fix bug when disable Account app freezes
- After login, pass App password as params for Add Dex Modal and Create Wallet Modal, then confirm registration will not ask for input password

Commit:
Update the Create Dex Modal UI with two options
- 1st tab show available servers to pick
- 2nd tab use for add custom server

Sirmorrison commented on ui/modal/create_password_modal.go:
```diff
@@ -148,6 +157,21 @@ func (cm *CreatePasswordModal) SetDescription(description string) *CreatePasswor
+func (cm *CreatePasswordModal) WithOldPasswordCreated(callback func(oldPassword, password string, m *CreatePasswordModal) bool) *CreatePasswordModal {
+	cm.withOldPasswordcallback = callback
+	return cm
+}
```
Why do we have an old password callback function? It is my believe that you want to implement a change password functionality. if this is true, then you dont need this.

Use the password modal to collect the old password, and if it checkout, use the create password modal for the new password.

this was how it was implemented on the settings page for start up wallet.

```
for pg.changeStartupPass.Clicked() {
		modal.NewPasswordModal(pg.Load).
			Title("Confirm current startup password").
			Hint("Current startup password").
			NegativeButton(values.String(values.StrCancel), func() {}).
			PositiveButton(values.String(values.StrConfirm), func(password string, pm *modal.PasswordModal) bool {
				go func() {
					var error string
					err := pg.wal.GetMultiWallet().VerifyStartupPassphrase([]byte(password))
					if err != nil {
						if err.Error() == dcrlibwallet.ErrInvalidPassphrase {
							error = "Invalid password"
						} else {
							error = err.Error()
						}
						pm.SetError(error)
						pm.SetLoading(false)
						return
					}
					pm.Dismiss()

					// change password
					modal.NewCreatePasswordModal(pg.Load).
						Title(values.String(values.StrCreateStartupPassword)).
						EnableName(false).
						PasswordHint("New startup password").
						ConfirmPasswordHint("Confirm new startup password").
						PasswordCreated(func(walletName, newPassword string, m *modal.CreatePasswordModal) bool {
							go func() {
								err := pg.wal.GetMultiWallet().ChangeStartupPassphrase([]byte(password), []byte(newPassword), dcrlibwallet.PassphraseTypePass)
								if err != nil {
									m.SetError(err.Error())
									m.SetLoading(false)
									return
								}
								pg.Toast.Notify("Startup password changed")
								m.Dismiss()
							}()
							return false
						}).Show()

				}()

				return false
			}).Show()
		break
	}
```

song50119 replied:
Have to confirm new password field is make sure user have correct one new, like almost app have.

song50119 replied:
Two  steps to do that? but ok now.

Sirmorrison replied:
We are planning add some security to the password modal reason why we have one for setting password and another for verifying it.

Sirmorrison commented on ui/page/dexclient/add_dex_modal.go:
```diff
@@ -121,33 +269,49 @@ func (md *addDexModal) Handle() {
+				}),
+				layout.Flexed(.5, func(gtx C) D {
+					return md.customServerBtn.Layout(gtx)
 				}),
```
```suggestion
				layout.Flexed(.5, md.customServerBtn.Layout),
```

Sirmorrison commented on ui/page/dexclient/add_dex_modal.go:
```diff
@@ -121,33 +269,49 @@ func (md *addDexModal) Handle() {
+			return layout.Flex{Spacing: layout.SpaceBetween, Alignment: layout.Middle}.Layout(gtx,
+				layout.Flexed(.5, func(gtx C) D {
+					return md.listServerBtn.Layout(gtx)
 				}),
```
```suggestion
			return layout.Flex{Spacing: layout.SpaceBetween, Alignment: layout.Middle}.Layout(gtx,
				layout.Flexed(.5, md.listServerBtn.Layout),
```

Sirmorrison commented on ui/page/dexclient/add_dex_modal.go:
```diff
@@ -177,35 +341,114 @@ func (md *addDexModal) Layout(gtx layout.Context) D {
+				return layout.Flex{Spacing: layout.SpaceBetween, Alignment: layout.Middle}.Layout(gtx,
+					layout.Rigid(func(gtx C) D {
+						return md.Theme.Label(values.MarginPadding14, md.listExchangeWdg[i].host).Layout(gtx)
+					}),
```
```suggestion
					layout.Rigid(md.Theme.Label(values.MarginPadding14, md.listExchangeWdg[i].host).Layout),
```

Sirmorrison commented on ui/page/dexclient/markets.go:
```diff
@@ -59,116 +72,187 @@ func (pg *Page) ID() string {
 
+func (pg *Page) headerLayout() layout.Widget {
+	return func(gtx C) D {
+		bottom := layout.Inset{Bottom: values.MarginPadding15}
```
this is not needed as it was only used once on line 152. It can fit into the same line as it was used. Kindly remove.

Sirmorrison commented on ui/page/dexclient/selector_dex_modal.go:
```diff
@@ -0,0 +1,133 @@
+						return layout.Flex{Spacing: layout.SpaceBetween, Alignment: layout.Middle}.Layout(gtx,
+							layout.Rigid(func(gtx C) D {
+								return md.Theme.Label(values.TextSize14, selectorExchangeWdg.Host).Layout(gtx)
+							}),
```
```suggestion
							layout.Rigid(md.Theme.Label(values.TextSize14, selectorExchangeWdg.Host).Layout),
```

Sirmorrison commented on ui/page/dexclient/settings_page.go:
```diff
@@ -0,0 +1,396 @@
+							})
+						})
+					})
+
```
Remove white space

Sirmorrison commented on ui/page/dexclient/settings_page.go:
```diff
@@ -0,0 +1,396 @@
+				return false
+			}).Show()
+	}
+}
```
Check my review comment above

Sirmorrison commented on ui/page/dexclient/utils.go:
```diff
@@ -0,0 +1,292 @@
+	if ord.Type == order.MarketOrderType {
+		return "market"
+	}
+	// return ord.Type == Market ? 'market' : Doc.formatCoinValue(ord.rate / 1e8)
```
Was this left on purpose? if not needed, kindly remove.

Sirmorrison commented on ui/page/dexclient/utils.go:
```diff
@@ -0,0 +1,292 @@
+		}
+	}
+
+	var settl float64 = 0
```
i believe you meant `settle`

Sirmorrison commented on ui/page/dexclient/wallet_deposit_modal.go:
```diff
@@ -0,0 +1,218 @@
+	}.Layout(gtx, func(gtx layout.Context) layout.Dimensions {
+		return layout.Flex{Alignment: layout.Middle}.Layout(gtx,
+			layout.Rigid(func(gtx C) D {
+				card.Radius = decredmaterial.CornerRadius{TopRight: 8, TopLeft: 8, BottomRight: 8, BottomLeft: 8}
```
```suggestion
				card.Radius = decredmaterial.Radius(8)
```

Commit:
Update Change App Password Modal
- Open Create New Password modal after input current password and handler errors
- Remove redundant comments

Commit:
Update confirm registration
- Add Asset Selector Modal and listing available asset to be able to select to pay the fee
- Display wallet status, confirmations number and amount have to pay the fee
- When click a row, check Asset not have wallet will show Create Wallet Modal otherwise will show Confirm Modal to finish register
- Format coin amount by Conventional Factor
- Change all variables named dex to dexServer
- Move hardcode strings to const variable

dreacot reviewed (changes requested):
The error message shown to the user can be simplified, while the longer version can be added to the logs
<img width="845" alt="Screenshot 2021-12-30 at 1 51 28 PM" src="https://user-images.githubusercontent.com/25265396/147753563-f77ea603-764a-47c3-a3f0-ebffecb53b3f.png">

While trying to change dex password, the inputted password should be validated on this modal, before taking you to the next modal
<img width="797" alt="Screenshot 2021-12-30 at 1 55 48 PM" src="https://user-images.githubusercontent.com/25265396/147753859-3a6d11b0-3264-4a5d-91c0-f74158bc215d.png">

Commit:
Fix crash when initialize mini trade form

Sirmorrison commented on ui/page/dexclient/add_dex_modal.go:
```diff
@@ -66,88 +103,202 @@ func (md *addDexModal) Dismiss() {
-						gtx.Constraints.Max.Y = 350
-						return md.cert.Layout(gtx)
-					})
+					return layout.Inset{
+						Left:  values.MarginPadding1,
+						Right: values.MarginPadding1,
+					}.Layout(gtx, func(gtx C) D { return D{} })
```
This seems redundant.

Sirmorrison commented on ui/page/dexclient/create_wallet_modal.go:
```diff
@@ -221,3 +241,259 @@ func (md *createWalletModal) Layout(gtx layout.Context) D {
+}
+
+func (amd *assetSelectorModal) OnDismiss() {
+}
```
```suggestion
func (amd *assetSelectorModal) OnDismiss() {}
```

Sirmorrison commented on ui/page/dexclient/create_wallet_modal.go:
```diff
@@ -221,3 +241,259 @@ func (md *createWalletModal) Layout(gtx layout.Context) D {
+		childrens := []layout.FlexChild{
+			layout.Rigid(func(gtx C) D {
+				return amd.Theme.Label(values.TextSize12, strMarket).Layout(gtx)
+			}),
```
```suggestion
			layout.Rigid(amd.Theme.Label(values.TextSize12, strMarket).Layout),
```

Sirmorrison commented on ui/page/dexclient/markets.go:
```diff
@@ -59,121 +73,181 @@ func (pg *Page) ID() string {
+				t := fmt.Sprintf("%d/%d", currentConfs, reqConfirms)
+				return pg.Theme.Label(values.TextSize14, t).Layout(gtx)
+			}),
+		)
```
```suggestion
	return func(gtx C) D {
	
	        txtLabel := func(gtx C, txt string)D{
	            return pg.Theme.Label(values.TextSize14, txt).Layout(gtx)
	        }
	
		reqConfirms, currentConfs := pg.dexServer.Fee.Confs, pg.dexServer.PendingFee.Confs
		return layout.Flex{Axis: layout.Vertical, Alignment: layout.Middle}.Layout(gtx,
			layout.Rigid(func(gtx C) D {
			        return txtLabel(gtx, strWaitingConfirms)
			}),
			layout.Rigid(func(gtx C) D {
				return txtLabel(gtx, fmt.Sprintf(nStrConfirmationsStatus, pg.dexServer.Host, reqConfirms))
			}),
			layout.Rigid(func(gtx C) D {
				return txtLabel(gtx, fmt.Sprintf("%d/%d", currentConfs, reqConfirms))
			}),
		)
```

Sirmorrison commented on ui/page/dexclient/mini_trade_form.go:
```diff
@@ -0,0 +1,217 @@
+		layout.Rigid(func(gtx C) D {
+			return layout.Inset{Top: values.MarginPadding10}.Layout(gtx, func(gtx C) D {
+				return m.submitBtn.Layout(gtx)
+			})
```
```suggestion
			return layout.Inset{Top: values.MarginPadding10}.Layout(gtx, m.submitBtn.Layout)
```

Commit:
Remove redundant codes

itswisdomagain commented on go.mod:
```diff
@@ -19,6 +19,7 @@ require (
 	github.com/gopherjs/gopherjs v0.0.0-20210901121439-eee08aaf2717 // indirect
 	github.com/jessevdk/go-flags v1.4.1-0.20200711081900-c17162fe8fd7
 	github.com/jrick/logrotate v1.0.0
+	github.com/ncruces/zenity v0.7.12
```
Do we really need this? Is it vetted? Can we do without it?

song50119 replied:
Godcr not yet support files system dialogs. This is require now

itswisdomagain commented on ui/page/dexclient/markets.go:
```diff
@@ -182,44 +253,83 @@ func (pg *Page) Handle() {
 					}
-					pg.Toast.Notify("App password created")
-
+					pg.Toast.Notify(strSuccessfully)
```
"Successfully" -> "Successful"

itswisdomagain commented on ui/page/dexclient/add_dex_modal.go:
```diff
@@ -21,34 +20,72 @@ import (
+		Right:  values.MarginPadding9,
 	}
+	md.listServerBtn.Inset, md.customServerBtn.Inset = inset, inset
+	md.listServerBtn.TextSize, md.customServerBtn.TextSize = values.TextSize14, values.TextSize14
```
I think this code will be easier to follow if you made it into a helper function that sets these fields for each button e.g.:
```
tabButton := func(text string, active bool) decredmaterial.Button {
	btn := l.Theme.OutlineButton(text)
	btn.CornerRadius = values.MarginPadding0
	btn.Inset = layout.Inset{
		Top:    values.MarginPadding5,
		Bottom: values.MarginPadding5,
		Left:   values.MarginPadding9,
		Right:  values.MarginPadding9,
	}
	btn.TextSize = values.TextSize14
	if !active {
		btn.Background = l.Theme.Color.Background
	}
	return btn
}
```

Then you can do
```
listServerBtn:    tabButton(strPickAServer, true),
customServerBtn:  tabButton(strCustomServer, false),
```

itswisdomagain commented on ui/page/dexclient/add_dex_modal.go:
```diff
@@ -21,34 +20,72 @@ import (
+	md.listServerBtn.TextSize, md.customServerBtn.TextSize = values.TextSize14, values.TextSize14
+
+	md.appPassword.Editor.SingleLine = true
+	md.dexServerAddress.Editor.SingleLine = true
```
I think it'll be nice to have a `l.Theme.SingleLineEditor` method instead of setting this separately for each editor field that requires it.

song50119 replied:
I see app's editor almost have singleline for input. How about set singleline by default? Then create method to enable multiple lines if needed?

itswisdomagain commented on ui/page/dexclient/add_dex_modal.go:
```diff
@@ -66,88 +103,202 @@ func (md *addDexModal) Dismiss() {
+
+	// Initialize listExchangeWdg.
+	certs := core.CertStore[md.Dexc().Core().Network()]
+	md.listExchangeWdg = make([]*knownExchangeWidget, 0)
```
```suggestion
	md.listExchangeWdg = make([]*knownExchangeWidget, 0, len(certs))
```

itswisdomagain commented on ui/page/dexclient/add_dex_modal.go:
```diff
@@ -66,88 +103,202 @@ func (md *addDexModal) Dismiss() {
 		}
 
-		md.isSending = true
 		go func() {
```
You should set `md.isSending` to true before starting the goroutine. If it's not set immediately, the goroutine may be delayed and the goroutine will be started twice if the user clicks this button twice.

itswisdomagain commented on ui/page/dexclient/add_dex_modal.go:
```diff
@@ -66,88 +103,202 @@ func (md *addDexModal) Dismiss() {
+
+			md.isSending = true
+			dexServer, paid, err := md.Dexc().Core().DiscoverAccount(serverAddr, []byte(appPass), cert)
 			md.isSending = false
```
Should only be false when this goroutine ends, to prevent duplicate actions. Use a defer at the beginning of the goroutine.

itswisdomagain commented on ui/page/dexclient/add_dex_modal.go:
```diff
@@ -66,88 +103,202 @@ func (md *addDexModal) Dismiss() {
+				Show:      md.Show,
+				completed: md.onDexCreated,
+				Dismiss:   md.Dismiss,
+			}
```
This can be moved into the `AssetSelected` callback, it's only used there.

itswisdomagain commented on ui/page/dexclient/add_dex_modal.go:
```diff
@@ -21,34 +20,72 @@ import (
+	appPassword  decredmaterial.Editor
+	onDexCreated func(*core.Exchange)
+
+	listServer        *widget.List
```
I believe we can do without this property, instead create a new `*widget.List` each time you want to layout the list. The UI elements whose state we want to keep track of here are just the individual buttons and that's tracked in `listExchangeWdg`.

itswisdomagain commented on ui/page/dexclient/add_dex_modal.go:
```diff
@@ -21,34 +20,72 @@ import (
+	onDexCreated func(*core.Exchange)
+
+	listServer        *widget.List
+	listExchangeWdg   []*knownExchangeWidget
```
I'll recommend a `map[string]*decredmaterial.Clickable` instead.

song50119 replied:
If use the map it will render random position of item on the UI

itswisdomagain commented on ui/page/dexclient/create_wallet_modal.go:
```diff
@@ -221,3 +239,256 @@ func (md *createWalletModal) Layout(gtx layout.Context) D {
+
+const DexassetSelectorModalID = "dex_asset_selector_modal"
+
+type assetSelectorModal struct {
```
Please move the `assetSelectorModal` and related code to a different file.

itswisdomagain commented on ui/page/dexclient/create_wallet_modal.go:
```diff
@@ -221,3 +239,256 @@ func (md *createWalletModal) Layout(gtx layout.Context) D {
 	return md.modal.Layout(gtx, w)
 }
+
+const DexassetSelectorModalID = "dex_asset_selector_modal"
```
```suggestion
const DexAssetSelectorModalID = "dex_asset_selector_modal"
```

itswisdomagain commented on ui/page/dexclient/add_dex_modal.go:
```diff
@@ -66,88 +103,202 @@ func (md *addDexModal) Dismiss() {
+	}
+}
+
+func (md *addDexModal) DexCreated(callback func(*core.Exchange)) *addDexModal {
```
This will probably read better as `OnRegistrationComplete` since we're not really creating a DEX, or possibly `OnDexAdded` since this is a `AddDexModal`.

itswisdomagain commented on ui/page/dexclient/create_wallet_modal.go:
```diff
@@ -221,3 +239,256 @@ func (md *createWalletModal) Layout(gtx layout.Context) D {
+
+type assetSelectorModal struct {
+	*load.Load
+	*core.Exchange
```
Any reason for making this an embedded field? It doesn't make sense that an `assetSelectorModal` is an extension of a `core.Exchange`.

itswisdomagain commented on ui/page/dexclient/create_wallet_modal.go:
```diff
@@ -221,3 +239,256 @@ func (md *createWalletModal) Layout(gtx layout.Context) D {
+type assetSelectWidget struct {
+	selectBtn *decredmaterial.Clickable
+	feeAsset  *core.FeeAsset
+	asset     *core.SupportedAsset
```
I'd say don't store this in memory. You can always access it later when needed using `amd.Dexc().Core().SupportedAssets()[feeAsset.ID]`.

itswisdomagain commented on ui/page/dexclient/create_wallet_modal.go:
```diff
@@ -221,3 +239,256 @@ func (md *createWalletModal) Layout(gtx layout.Context) D {
+	listWalletWdg []*assetSelectWidget
+	modal         *decredmaterial.Modal
+	cancelBtn     decredmaterial.Button
+	assetSelected func(*core.SupportedAsset)
```
`onAssetSelected`

itswisdomagain commented on ui/page/dexclient/create_wallet_modal.go:
```diff
@@ -221,3 +239,256 @@ func (md *createWalletModal) Layout(gtx layout.Context) D {
+type assetSelectorModal struct {
+	*load.Load
+	*core.Exchange
+	listWallet    *widget.List
```
Don't think this is necessary. See a similar comment on the `add_dex_modal` file.

itswisdomagain commented on ui/page/dexclient/create_wallet_modal.go:
```diff
@@ -221,3 +239,256 @@ func (md *createWalletModal) Layout(gtx layout.Context) D {
+
+func (amd *assetSelectorModal) OnDismiss() {}
+
+func (amd *assetSelectorModal) AssetSelected(callback func(*core.SupportedAsset)) *assetSelectorModal {
```
OnAssetSelected

itswisdomagain reviewed (changes requested):
Haven't gone through all the changes but the few changes I've seen look really good. Good job! This review only covers the initial steps of using the DEX feature i.e. the modals for adding a dex and selecting an asset to use to pay the fee. Will comment on the next steps in subsequent review(s).

Something to keep in mind: use as few in-memory variables as is possible. The more variables kept in memory, the more memory-intensive and heavier the app becomes and could become slow on systems with limited memory. If it's possible to get a particular data by calling a method/function repeatedly, do so rather than calling the method once and saving the result in a variable to be used later. That is, for functions and methods that are inexpensive.

Commit:
Validate input invoice amout
- Display rate step and lot size at on UI
- Adjust spacing left-right of Dex Settig and Order History page
- Calculate wallet sync progress and display at mini trade form
- Fix lag/delay in the app when opening the DEX page for the first time
- Move Asset Selector Modal to new file

Commit:
Validate inputs before submit Add New Dex
- Validate input before adding new wallet
- Add SingleLineEditor method to theme

Commit:
Refactor Settings Page
- change exchangesWdg field to accountHandlerBtns
- change settingExchangeWidget to accountHandlerButton struct

Commit:
call ctxCancel at OnNavigatedFrom Market Page

Commit:
Update Gio version

itswisdomagain commented on go.mod:
```diff
@@ -4,7 +4,8 @@ go 1.16
 	decred.org/dcrdex v0.4.0-rc3
-	gioui.org v0.0.0-20211011183043-05f0f5c20f45
+	gioui.org v0.0.0-20220105104929-8d8aeef66bef
+	gioui.org/x/explorer v0.0.0-20220105205654-261c7273d121
```
[The doc](https://pkg.go.dev/gioui.org/x/explorer#readme-status) say `Currently, Explorer supports across Android, JS and Windows 10. It will return ErrAvailableAPI for any other platform that isn't supported.` How do we feel about this?

song50119 replied:
Docs might not update yet, i can use on MacOs

itswisdomagain commented on ui/page/dexclient/add_dex_modal.go:
```diff
@@ -21,34 +19,66 @@ import (
+	cancelBtn        decredmaterial.Button
 	materialLoader   material.LoaderStyle
+	certContent      []byte
+	fileSelectBtn    *decredmaterial.Clickable
```
`selectCertBtn`

itswisdomagain commented on ui/page/dexclient/add_dex_modal.go:
```diff
@@ -21,34 +19,66 @@ import (
+	certContent      []byte
+	fileSelectBtn    *decredmaterial.Clickable
+
+	// defaultAppPass is the password value after login or initialize to continue processing add new DEX
+	// the Add Dex Modal won't show password input on UI
+	defaultAppPass string
+	appPassword    decredmaterial.Editor
```
```suggestion
	// appPass is set after login or initialize to continue process of adding a DEX
	// without requesting password input on UI
	appPass string
	appPasswordEditor    decredmaterial.Editor
```

itswisdomagain commented on ui/page/dexclient/add_dex_modal.go:
```diff
@@ -66,91 +96,222 @@ func (md *addDexModal) Dismiss() {
+		return cl
+	}
+
+	// Initialize listExchangeWdg.
```
// Prepare buttons for selecting pre-defined exchanges.

itswisdomagain commented on ui/page/dexclient/add_dex_modal.go:
```diff
@@ -66,91 +96,222 @@ func (md *addDexModal) Dismiss() {
+	}
+
+	// Initialize listExchangeWdg.
+	listServer := sliceSever(core.CertStore[md.Dexc().Core().Network()])
```
Please rename `sliceSever` to `sortExchanges`.
```suggestion
	dexServers := sortExchanges(core.CertStore[md.Dexc().Core().Network()])
```

itswisdomagain reviewed (changes requested):
Stopping review here to recommend splitting this PR into multiple PRs for easier review. Can keep this PR open but open a separate PR with just the first few commits in this PR. Once that's reviewed and merged, you can rebase this PR to remove those commits and this can be reviewed again.
